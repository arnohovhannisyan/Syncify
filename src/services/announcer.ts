import { window, commands, env, Uri } from "vscode";
import { localize } from "~/services";

export async function showAnnouncement(url: string): Promise<void> {
	return new Promise(resolve => {
		const message = window.setStatusBarMessage(
			localize("(info) announcementAvailable")
		);

		const openBtn = window.createStatusBarItem(1);
		const dismissBtn = window.createStatusBarItem(1);

		const dispose = () => {
			openDisposable.dispose();
			dismissDisposable.dispose();
			openBtn.dispose();
			dismissBtn.dispose();
			message.dispose();

			resolve();
		};

		const openDisposable = commands.registerCommand(
			"syncify.openAnnouncement",
			() => {
				env.openExternal(Uri.parse(url));

				dispose();
			}
		);

		const dismissDisposable = commands.registerCommand(
			"syncify.dismissAnnouncement",
			dispose
		);

		openBtn.command = "syncify.openAnnouncement";
		openBtn.text = `$(check) ${localize("(label) open")}`;
		openBtn.show();

		dismissBtn.command = "syncify.dismissAnnouncement";
		dismissBtn.text = `$(x) ${localize("(label) dismiss")}`;
		dismissBtn.show();
	});
}
