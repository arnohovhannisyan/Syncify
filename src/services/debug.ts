import { commands } from "vscode";
import actions from "~/redux/actions";
import { store } from "~/redux/store";

export class Debug {
  public static log(...args: any[]) {
    const { isDebugMode } = store.getState();

    if (isDebugMode) console.log(...args);
  }

  public static setDebug(value: boolean) {
    if (value) {
      store.dispatch(actions.enableDebugMode());
    } else {
      store.dispatch(actions.disableDebugMode());
    }

    return commands.executeCommand("syncify.reinitialize");
  }
}
