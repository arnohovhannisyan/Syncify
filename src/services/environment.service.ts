import { normalize, resolve } from "path";
import { state } from "../state";

export class EnvironmentService {
  public locations = {
    userFolder: null
  };

  constructor() {
    const slash = normalize("/");
    if (!process.env.VSCODE_PORTABLE) {
      const path = resolve(state.context.globalStoragePath, "../../..").concat(
        slash
      );
      this.locations.userFolder = resolve(path, "User").concat(slash);
    } else {
      const path = process.env.VSCODE_PORTABLE;
      this.locations.userFolder = resolve(path, "user-data/User").concat(slash);
    }
  }
}
