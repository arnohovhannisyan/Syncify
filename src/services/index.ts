export * from "~/services/environment.service";
export * from "~/services/localization.service";
export * from "~/services/custom-file.service";
export * from "~/services/extensions.service";
export * from "~/services/factory.service";
export * from "~/services/fs.service";
export * from "~/services/oauth.service";
export * from "~/services/init.service";
export * from "~/services/logger.service";
export * from "~/services/pragma.service";
export * from "~/services/settings.service";
export * from "~/services/utilities.service";
export * from "~/services/watcher.service";
export * from "~/services/webview.service";
