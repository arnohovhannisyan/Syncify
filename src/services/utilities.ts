import cloneDeep from "lodash/cloneDeep";
import merge from "lodash/merge";

export class Utilities {
  public static sleep(ms: number): Promise<void> {
    return new Promise(res => setTimeout(res, ms));
  }

  public static merge<TObj, TSrc>(object: TObj, source: TSrc): TObj & TSrc {
    return merge<TObj, TSrc>(cloneDeep(object), source);
  }
}
