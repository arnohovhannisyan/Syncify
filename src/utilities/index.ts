export * from "~/utilities/isJSON";
export * from "~/utilities/confirm";
export * from "~/utilities/sleep";
export * from "~/utilities/merge";
export * from "~/utilities/checkGit";
export * from "~/utilities/stringifyPretty";
