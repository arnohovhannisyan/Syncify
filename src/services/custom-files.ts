import { Environment, FS, localize, Logger } from "~/services";
import { basename, resolve } from "path";
import {
	InputBoxOptions,
	OpenDialogOptions,
	QuickPickItem,
	Uri,
	window,
	workspace,
} from "vscode";

export const importFile = async (uri?: Uri): Promise<void> => {
	try {
		const customFilesFolder = await Environment.customFilesFolder();

		await FS.mkdir(customFilesFolder);

		const allFiles = await FS.listFiles(customFilesFolder, ["**/*"]);

		if (allFiles.length === 0) {
			await window.showInformationMessage(
				localize("(info) CustomFiles.import -> noFiles"),
			);

			return;
		}

		const folder = await (async () => {
			if (uri) return uri.fsPath;

			if (!workspace.workspaceFolders) return;

			if (workspace.workspaceFolders.length === 1) {
				return workspace.workspaceFolders[0].uri.fsPath;
			}

			const result = await window.showQuickPick(
				workspace.workspaceFolders.map<QuickPickItem>((folder) => ({
					label: folder.name,
					description: folder.uri.fsPath,
				})),
				{
					placeHolder: localize(
						"(prompt) CustomFiles.import -> folder placeholder",
					),
				},
			);

			if (!result) return;

			const selectedWorkspace = workspace.workspaceFolders.find(
				(folder) => folder.uri.fsPath === result.description,
			);

			return selectedWorkspace?.uri.fsPath;
		})();

		if (!folder) return;

		const selectedFile = await window.showQuickPick(
			allFiles.map((file) => basename(file)),
			{
				placeHolder: localize(
					"(prompt) CustomFiles.import -> file placeholder",
				),
			},
		);

		if (!selectedFile) return;

		const filepath = resolve(customFilesFolder, selectedFile);

		const options: InputBoxOptions = {
			prompt: localize("(prompt) CustomFiles.import -> file name"),
			value: selectedFile,
		};

		const filename = (await window.showInputBox(options)) ?? selectedFile;

		if (filename.length === 0) return;

		const contents = await FS.readBuffer(filepath);
		await FS.write(resolve(folder, filename), contents);
	} catch (error) {
		void Logger.error(error);
	}
};

const registerDialogOptions: OpenDialogOptions = {
	canSelectMany: false,
	openLabel: localize("(label) CustomFiles.register -> select file"),
};

export const registerFile = async (uri?: Uri): Promise<void> => {
	try {
		const customFilesFolder = await Environment.customFilesFolder();

		await FS.mkdir(customFilesFolder);

		const filepath =
			uri?.fsPath ??
			(await window.showOpenDialog(registerDialogOptions))?.[0]?.fsPath;

		if (!filepath) return;

		const originalFilename = basename(filepath);

		const options: InputBoxOptions = {
			prompt: localize("(prompt) CustomFiles.register -> name"),
			value: originalFilename,
		};

		const filename = (await window.showInputBox(options)) ?? originalFilename;

		if (filename.length === 0) return;

		const newPath = resolve(customFilesFolder, filename);

		if (await FS.exists(newPath)) {
			const result = await window.showWarningMessage(
				localize("(prompt) CustomFiles.register -> file exists"),
				localize("(label) no"),
				localize("(label) yes"),
			);

			if (result !== localize("(label) yes")) return;
		}

		const contents = await FS.readBuffer(filepath);

		await FS.write(newPath, contents);

		await window.showInformationMessage(
			localize("(info) CustomFiles.register -> registered", filename),
		);
	} catch (error) {
		void Logger.error(error);
	}
};
