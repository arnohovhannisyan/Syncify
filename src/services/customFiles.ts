import { basename, resolve } from "path";
import { Uri, window, workspace } from "vscode";
import { Environment, FS, localize } from "~/services";

export class CustomFiles {
  public static async import(uri?: Uri): Promise<void> {
    if (!workspace.workspaceFolders.length) {
      return;
    }

    const folderExists = await FS.exists(Environment.customFilesFolder);

    if (!folderExists) {
      await FS.mkdir(Environment.customFilesFolder);
    }

    const allFiles = await FS.listFiles(Environment.customFilesFolder);

    if (!allFiles.length) {
      await window.showInformationMessage(
        localize("(info) customFiles.noFilesAvailable")
      );
      return;
    }

    const folder = await (async () => {
      if (uri) {
        return uri.fsPath;
      }

      if (workspace.workspaceFolders.length === 1) {
        return workspace.workspaceFolders[0].uri.fsPath;
      }

      const result = await window.showQuickPick(
        workspace.workspaceFolders.map(f => f.name),
        {
          placeHolder: localize("(prompt) customFile.import.folderPlaceholder")
        }
      );

      return workspace.workspaceFolders.filter(f => f.name === result)[0].uri
        .fsPath;
    })();

    if (!folder) {
      return;
    }

    const filename = await window.showQuickPick(
      allFiles.map(f => basename(f)),
      {
        placeHolder: localize("(prompt) customFile.import.filePlaceholder")
      }
    );

    if (!filename) {
      return;
    }

    const filepath = resolve(Environment.customFilesFolder, filename);
    const contents = await FS.readBuffer(filepath);
    await FS.write(resolve(folder, filename), contents);
  }

  public static async register(uri: Uri) {
    const folderExists = await FS.exists(Environment.customFilesFolder);

    if (!folderExists) {
      await FS.mkdir(Environment.customFilesFolder);
    }

    const filepath = uri
      ? uri.fsPath
      : await (async () => {
          const result = await window.showOpenDialog({
            canSelectMany: false,
            openLabel: localize("(action) customFiles.selectFile")
          });

          if (!result) {
            return;
          }

          return result[0].fsPath;
        })();

    if (!filepath) {
      return;
    }

    const filename = basename(filepath);
    const contents = await FS.readBuffer(filepath);
    const newPath = resolve(Environment.customFilesFolder, filename);
    await FS.write(newPath, contents);
    await window.showInformationMessage(
      localize("(info) customFiles.registered", filename)
    );
  }
}
