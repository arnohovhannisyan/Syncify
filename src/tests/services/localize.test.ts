import { initLocalization, localize, FS, Environment } from "~/services";
import { resolve } from "path";
import { getCleanupPath } from "~/tests/getCleanupPath";
import { stringifyPretty } from "~/utilities";

test("language detection", async () => {
	const spy = jest.spyOn(FS, "exists");

	{
		const spy = jest.spyOn(FS, "exists");

		process.env.VSCODE_NLS_CONFIG = JSON.stringify({ locale: "fake-locale" });

		await initLocalization();

		delete process.env.VSCODE_NLS_CONFIG;

		const expected = resolve(
			Environment.extensionPath,
			`package.nls.fake-locale.json`
		);

		expect(spy).toHaveBeenCalledWith(expected);

		spy.mockClear();
	}

	{
		await initLocalization();

		const expected = resolve(
			Environment.extensionPath,
			`package.nls.en-us.json`
		);

		expect(spy).toHaveBeenCalledWith(expected);

		spy.mockClear();
	}

	spy.mockRestore();
});

test("returns requested language pack", async () => {
	const cleanupPath = getCleanupPath("services/localize");

	const spy = jest
		.spyOn(Environment, "extensionPath", "get")
		.mockReturnValueOnce(cleanupPath);

	await FS.mkdir(cleanupPath);

	await FS.write(
		resolve(cleanupPath, "package.nls.lang.json"),
		stringifyPretty({
			key: "value"
		})
	);

	await initLocalization("lang");

	expect(localize("key")).toBe("value");

	await FS.remove(cleanupPath);

	spy.mockRestore();
});

test("basic functionality", async () => {
	await initLocalization("en-us");

	expect(localize("(info) extensions -> installed", "5")).toMatchSnapshot();
	expect(localize("(info) extensions -> uninstalled", "10")).toMatchSnapshot();
});

test("invalid key", async () => {
	await initLocalization("en-us");

	expect(localize("")).toBe("");

	const rand = Math.random().toString();

	expect(localize(rand)).toBe(rand);
});
