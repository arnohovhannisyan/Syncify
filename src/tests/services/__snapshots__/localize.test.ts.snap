// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic functionality 1`] = `"Syncify: Installed 5"`;

exports[`basic functionality 2`] = `"Syncify: Uninstalled 10"`;
